sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java

```diff
@@ -0,0 +1,557 @@
+     *     .password("semp-password")
+     *     .vpnName("vpn-name")
+     *     .build()));
+     * }</pre>
+     */
+    public Read<T> withSessionServiceFactory(SessionServiceFactory sessionServiceFactory) {
+      checkState(
+          sessionServiceFactory != null, "SolaceIO.Read: sessionServiceFactory must not be null.");
+      return toBuilder().setSessionServiceFactory(sessionServiceFactory).build();
+    }
+
+    abstract @Nullable Queue getQueue();
+
+    abstract @Nullable Topic getTopic();
+
+    abstract @Nullable SerializableFunction<T, Instant> getTimestampFn();
+
+    abstract @Nullable Integer getMaxNumConnections();
+
+    abstract boolean getDeduplicateRecords();
+
+    abstract SerializableFunction<@Nullable BytesXMLMessage, @Nullable T> getParseFn();
+
+    abstract @Nullable SempClientFactory getSempClientFactory();
+
+    abstract @Nullable SessionServiceFactory getSessionServiceFactory();
+
+    abstract TypeDescriptor<T> getTypeDescriptor();
+
+    public static <T> Builder<T> builder() {
+      Builder<T> builder = new org.apache.beam.sdk.io.solace.AutoValue_SolaceIO_Read.Builder<T>();
+      builder.setDeduplicateRecords(DEFAULT_DEDUPLICATE_RECORDS);
+      return builder;
+    }
+
+    abstract Builder<T> toBuilder();
+
+    @AutoValue.Builder
+    public abstract static class Builder<T> {
+
+      abstract Builder<T> setQueue(Queue queue);
+
+      abstract Builder<T> setTopic(Topic topic);
+
+      abstract Builder<T> setTimestampFn(SerializableFunction<T, Instant> timestampFn);
+
+      abstract Builder<T> setMaxNumConnections(Integer maxNumConnections);
+
+      abstract Builder<T> setDeduplicateRecords(boolean deduplicateRecords);
+
+      abstract Builder<T> setParseFn(
+          SerializableFunction<@Nullable BytesXMLMessage, @Nullable T> parseFn);
+
+      abstract Builder<T> setSempClientFactory(SempClientFactory brokerServiceFactory);
+
+      abstract Builder<T> setSessionServiceFactory(SessionServiceFactory sessionServiceFactory);
+
+      abstract Builder<T> setTypeDescriptor(TypeDescriptor<T> typeDescriptor);
+
+      abstract Read<T> build();
+    }
+
+    @Override
+    public PCollection<T> expand(PBegin input) {
+      checkState(
+          (getQueue() == null ^ getTopic() == null),
+          "SolaceIO.Read: One of the Solace {Queue, Topic} must be set.");
+
+      SempClientFactory sempClientFactory =
+          checkNotNull(getSempClientFactory(), "SolaceIO: sempClientFactory is null.");
+      String jobName = input.getPipeline().getOptions().getJobName();
+      Queue queueFromOptions = getQueue();
+      Queue initializedQueue =
+          queueFromOptions != null
+              ? queueFromOptions
+              : initializeQueueForTopic(jobName, sempClientFactory);
+
+      SessionServiceFactory sessionServiceFactory =
+          checkNotNull(getSessionServiceFactory(), "SolaceIO: sessionServiceFactory is null.");
+      sessionServiceFactory.setQueue(initializedQueue);
```

Comment:
Thanks for the suggestion!
Here's what I tried, using the `getSempClientFactory()` as an example. 
I moved the check to the `validate()` method
```
public void validate(@Nullable PipelineOptions options) {
      checkNotNull(getSempClientFactory(), "SolaceIO: sempClientFactory is null.");
    }
```
but I still need to do the extra validation in the expand() method, to satisfy the checkerframework checks.
```
SempClientFactory sempClientFactory = checkNotNull(getSempClientFactory());
```
Would that be ok, to duplicate the `checkNotNull` method call for each `@Nullable` method called in the `expand()`? 
Otherwise, I'd need to annotate `initializeQueueForTopic()` method arguments as `@Nullable`, 'dragging' this annotation unnecessarily downstream.
Unfortunately, I can't make the `getSempClientFactory()` return a NonNull value, it has to be Nullable, as I can't default to any sensible value.

Reply:
@bzablocki I will send a meeting request that we can summarize our discussion as a comment in this PR for others benefit.